lib/connection_pool.rb

```diff
@@ -53,22 +53,28 @@ def initialize(options = {}, &block)
   end
 
   def with(options = {})
+    success = false # hoisted
     conn = checkout(options)
     begin
-      yield conn
+      (yield conn).tap do
+        success = true # means the connection wasn't interrupted
+      end
     ensure
-      checkin
+      if success
+        # everything is roses, we can safely check the connection back in
+        checkin
+      else
+        @available.discard!(pop_connection)
```

Comment:
Thanks @tamird that helps.

Do you know why `checkin` and `@available.discard!(pop_connection)` would have a different wait behavior between threads then? I'm trying to explore to make sense of it, but I'm not understanding it.

I even tried this to see if I could force a block in in the `discard` branch, but it still failed.

```ruby
class CloseableMysqlConnection < ::BasicObject
  METHODS = [:close]

  def initialize(connection)
    @connection = connection
  end

  def close
    sleep 4 # just to see if this would work.
    @connection.close
  end

  def respond_to?(id, *args)
    METHODS.include?(id) || @connection.respond_to?(id, *args)
  end

  def method_missing(name, *args, &block)
    @connection.send(name, *args, &block)
  end
end


class ActiveRecord::Base
  def self.connection
    @@shared_connection || ConnectionPool::Wrapper.new(:size => 1) { CloseableMysqlConnection.new(retrieve_connection) }
  end
end
```

I'm having a heck of a time trying to put together a small lightweight example using threads to show the difference in behavior.